modules/standard/GMP.chpl

```diff
@@ -581,838 +602,993 @@ module GMP {
       var dbl:c_double;
-      on this {
+      on Locales[this.locale_id] {
         var tmp:c_long;
         dbl = mpz_get_d_2exp(tmp, this.mpz);
         exp = tmp;
       }
-      return (exp.safeCast(int), dbl: real);
+      return (exp.safeCast(uint(32)), dbl: real);
     }
-    proc get_str(base:int=10):string
+    proc get_str(base: int=10):string
     {
       var ret:string;
-      on this {
-        var tmp = chpl_gmp_mpz_get_str(base.safeCast(c_int), this.mpz);
-        ret = tmp:string;
+      on Locales[this.locale_id] {
+        var tmpvar = chpl_gmp_mpz_get_str(base.safeCast(c_int), this.mpz);
+        ret = new string(tmpvar, owned=true, needToCopy=false);
       }
       return ret;
     }
 
     // Arithmetic functions
-    proc add(a:BigInt, b:BigInt)
+    proc add(ref a: BigInt, ref b: BigInt)
     {
-      on this {
-        var (acopy,a_) = a.maybeCopy();
-        var (bcopy,b_) = b.maybeCopy();
-        mpz_add(this.mpz, a_.mpz, b_.mpz);
-        if acopy then delete a_;
-        if bcopy then delete b_;
+      on Locales[this.locale_id] {
+        if (here.id != a.locale_id || here.id != b.locale_id) {
+          var a_ = a;
+          var b_ = b;
+          mpz_add(this.mpz, a_.mpz, b_.mpz);
+        } else {
+          mpz_add(this.mpz, a.mpz, b.mpz);
+        }
       }
     }
-    proc add_ui(a:BigInt, b:uint)
+    proc add_ui(ref a: BigInt, b: uint)
     {
-      on this {
-        var (acopy,a_) = a.maybeCopy();
-        mpz_add_ui(this.mpz, a_.mpz, b.safeCast(c_ulong));
-        if acopy then delete a_;
+      on Locales[this.locale_id] {
+        if (here.id != a.locale_id) {
+          var a_ = a;
+          mpz_add_ui(this.mpz, a_.mpz, b.safeCast(c_ulong));
+        } else {
+          mpz_add_ui(this.mpz, a.mpz, b.safeCast(c_ulong));
+        }
       }
     }
-    proc sub(a:BigInt, b:BigInt)
+    proc sub(ref a: BigInt, ref b: BigInt)
     {
-      on this {
-        var (acopy,a_) = a.maybeCopy();
-        var (bcopy,b_) = b.maybeCopy();
-        mpz_sub(this.mpz, a_.mpz, b_.mpz);
-        if acopy then delete a_;
-        if bcopy then delete b_;
+      on Locales[this.locale_id] {
+        if (here.id != a.locale_id || here.id != b.locale_id) {
+          var a_ = a;
+          var b_ = b;
+          mpz_sub(this.mpz, a_.mpz, b_.mpz);
+        } else {
+          mpz_sub(this.mpz, a.mpz, b.mpz);
+        }
       }
     }
-    proc sub_ui(a:BigInt, b:uint)
+    proc sub_ui(ref a: BigInt, b: uint)
     {
-      on this {
-        var (acopy,a_) = a.maybeCopy();
-        mpz_sub_ui(this.mpz, a_.mpz, b.safeCast(c_ulong));
-        if acopy then delete a_;
+      on Locales[this.locale_id] {
+        if (here.id != a.locale_id) {
+          var a_ = a;
+          mpz_sub_ui(this.mpz, a_.mpz, b.safeCast(c_ulong));
+        } else {
+          mpz_sub_ui(this.mpz, a.mpz, b.safeCast(c_ulong));
+        }
       }
     }
-    proc ui_sub(a:uint, b:BigInt)
+    proc ui_sub(a: uint, ref b: BigInt)
```

Comment:
For face-to-face

I don't have any reason to believe there's a problem here but I'd like to work through the
mechanisms that enable the case for here.id != b.locale_id to function correctly i.e. the
deep copy required, any tests, etc.